### PR TITLE
Add `dir` field to raw, and handle it.

### LINKED ISSUE #3689: Syntax highlighted raw texts within RTL incorrect direction
### Description

```typ
// Mixing raw
#set text(lang: "he")
#show raw: set text(lang: "en", dir: ltr)
לדוג. `if a == b:` זה תנאי

#set raw(lang:"python")
לדוג. `if a == b:` זה תנאי
```

image

This is due to my PR #3532, since now each section of text is treated individually.

This possible problem was pointed out, but we didn't have an example of where it matters, now we do :laughing: 

> On another note: As implemented, this will individually isolate consecutive segments of text going against the paragraph direction (multiple segments can be forced e.g. by just putting `/**/` between words). Could that be a problem?

_Originally posted by @laurmaedje in https://github.com/typst/typst/issues/3532#issuecomment-1976050396_

Before #3532 for reference (notice the colon on the wrong side):
image

# Possible solutions
- Revert #3532, re-introducing the bug it was trying to solve (code within RTL text will still be broken)
- Add `dir` field to raw, just like text has, and properly handle raw text direction (would adding `lang` as well be necessary, I don't see why but maybe I'm missing something)
- Something completely different?

@laurmaedje would love your input on this, thanks!

### Reproduction URL

_No response_

### Operating system

_No response_

### Typst version

- [X] I am using the latest version of Typst

### PR DESCRIPTION
**Pull request by A-Walrus**
Closes #3689.

### CODE REVIEW

File: crates/typst/src/text/raw.rs [hunk 1]
```diff
@@ -263,6 +263,33 @@ pub struct RawElem {
...
+    /// output, please do get in touch with us through the
+    /// [contact form](https://typst.app/contact) or our
+    /// [Discord server]($community/#discord)!
+    ///
+    /// ```example
+    /// #set text(dir: rtl)
+    /// هذا عربي.
+    /// ```
+    #[resolve]
+    pub dir: TextDir,
```
Comment (A-Walrus) [inline crates/typst/src/text/raw.rs:292]: Copied this whole chunk from the text element, but removed the `ghost` attribute. Not sure if that causes an issue.

File: crates/typst/src/text/raw.rs [hunk 2]
```diff
@@ -433,6 +460,15 @@ impl Show for Packed<RawElem> {
         let lines = self.lines().map(|v| v.as_slice()).unwrap_or_default();
 
         let mut seq = EcoVec::with_capacity((2 * lines.len()).saturating_sub(1));
+
+        let dir = RawElem::dir_in(styles);
+        // Insert "Explicit Directional Isolate".
+        match dir {
+            Dir::LTR => seq.push(TextElem::new("\u{2066}".into()).pack()),
+            Dir::RTL => seq.push(TextElem::new("\u{2067}".into()).pack()),
+            _ => {}
```
Comment (A-Walrus) [inline crates/typst/src/text/raw.rs:469]: Should we do something with `TTB` and `BTT` (likewise for my previous changes to `inline/mod.rs` #3532)

**Comment by PgBiel** [comment]
I believe the idiomatic approach wouldn't be to introduce a `dir` parameter to `raw`, but rather override the `TextDir` in `impl ShowSet` by default and then have users optionally `#show raw: set text(dir: ...)` to change it.

**Comment by laurmaedje** [comment]
I agree with @PgBiel that a `dir` field on raw wouldn't be idiomatic. It seems to me that the proper fix would be to isolate consecutive segments of text going against the dominant direction individually in paragraph layout. If necessary, we can also add a default show set rule on raw blocks for LTR.

**Pull request closed without merging**